src/label/repelLabels.ts

```diff
@@ -0,0 +1,370 @@
+                        overlapCount[i] += 1;
+                        f.add(repelForce(
+                            ci, DataPoints[j],
+                            forcePush * forcePointSize * 0.003
+                        ));
+                    }
+                }
+                else {
+                    cj.set(LabelRects[j].x + LabelRects[j].width / 2, LabelRects[j].y + LabelRects[j].height / 2);
+                    // Repel the box from other data points.
+                    if (overlapsRect2(labelList[j].hostRect, LabelRects[i])) {
+                        totalOverlaps += 1;
+                        iOverlaps = true;
+                        overlapCount[i] += 1;
+                        f.add(repelForce(ci, DataPoints[j],
+                            forcePush * forcePointSize * 0.003
+                        ));
+                    }
+                    if (overlapsRect2(LabelRects[i], LabelRects[j])) {
+                        totalOverlaps += 1;
+                        iOverlaps = true;
+                        overlapCount[i] += 1;
+                        f.add(repelForce(ci, cj, forcePush));
+                    }
+                    if (iter % 100 === 0 && overlapCount[i] === 2
+                        && overlapsRect2(LabelRects[i], labelList[i].hostRect)
+                        && overlapsRect2(LabelRects[i], labelList[j].hostRect)) {
+                        const v = new Point();
+                        const stuck = isStuck(ci, DataPoints[i], DataPoints[j], v);
+                        if (stuck) {
+                            v.scale(10);
+                            if (Math.random() < 0.5) {
+                                v.set(v.x, -v.y);
+                                moveRect(LabelRects[i], Labels[i], v);
+                            }
+                            else {
+                                v.set(-v.x, v.y);
+                                moveRect(LabelRects[i], Labels[i], v);
+                            }
+                        }
+                    }
+                }
+            }
+
+            // Pull the box.
+            if (!iOverlaps) {
+                f.add(springForce(DataPoints[i], ci, forcePull * forcePointSize));
+            }
+            let friction2 = 1.0;
+            if (overlapCount[i] > 10) {
+                friction2 += 0.5;
+            }
+            else {
+                friction2 += 0.05 * overlapCount[i];
+            }
+            velocities[i].scale(friction * friction2);
+            velocities[i].add(f);
+            moveRect(labelList[i].rect, Labels[i], velocities[i]);
+
+            // Put boxes within bounds
+            putWithinBounds(LabelRects[i], Labels[i], xBounds, yBounds);
+
+            // check line intersect
+            if (totalOverlaps === 0 || iter % 10 === 0) {
+                for (let j = 0; j < labelList.length; j++) {
+                    ci.set(labelList[i].rect.x + labelList[i].rect.width / 2,

```

Comment:
It's better to assign `labelList[i].rect` to a temporal variable for smaller code size.